src/Symfony/Bundle/WebProfilerBundle/Tests/Controller/ProfilerControllerTest.php

```diff
@@ -185,17 +193,105 @@ public function provideCspVariants()
         ];
     }
 
-    private function createController($profiler, $twig, $withCSP): ProfilerController
+    /**
+     * @dataProvider bestPanelProvider
+     */
+    public function testBestPanel(string $expectedPanel, Profile $profile)
+    {
+        $profiler = $this->createMock(Profiler::class);
+        $profiler
+            ->expects($this->atLeastOnce())
+            ->method('loadProfile')
+            ->with($profile->getToken())
+            ->willReturn($profile);
+
+        $profiler
+            ->expects($this->atLeastOnce())
+            ->method('has')
+            ->with($this->logicalXor($collectorsNames = array_keys($profile->getCollectors())))
+            ->willReturn(true);
+
+        $loader = $this->createMock(SourceContextLoaderInterface::class);
+        $loader
+            ->expects($this->atLeastOnce())
+            ->method('getSourceContext')
+            ->with($this->logicalXor($expectedTemplate = 'expected_template.html.twig', 'other_template.html.twig'))
+            ->willReturn(new Source('foo', 'bar'));
+
+        $twig = $this->createMock(Environment::class);
+        $twig
+            ->expects($this->atLeastOnce())
+            ->method('getLoader')
+            ->willReturn($loader);
+        $twig
+            ->expects($this->once())
+            ->method('render')
+            ->with($expectedTemplate);
+
+        $request = new Request();
+        $request->query->set('panel', '_best');
+
+        $this
+            ->createController($profiler, $twig, false, array_map(function (string $collectorName) use ($expectedPanel, $expectedTemplate): array {
+                if ($collectorName === $expectedPanel) {
+                    return [$expectedPanel, $expectedTemplate];
+                }
+
+                return [$collectorName, 'other_template.html.twig'];
+            }, $collectorsNames))
+            ->panelAction($request, $profile->getToken());
+    }
+
+    public function bestPanelProvider(): \Generator
+    {
+        $profile = new Profile('xxxxxx');
+
+        // Test default behavior
+        $profile->addCollector($requestDataCollector = new RequestDataCollector());
+        yield [$requestDataCollector->getName(), $profile];
+
+        // Test exception
+        $profile = clone $profile;
```

Comment:
I think the provider will be much more readable if you create a new profile each time instead of cloning the previous one (so inheriting its state) and having to figure out what gets overridden.